File: Data Structures/BST/bst.py
class Node:
	"""docstring for Node"""
	def __init__(self, val, lchild, rchild):
		self.val = val
		self.lchild = lchild
		self.rchild = rchild
	
	def insert(self, value):
		if self.val < value:
			if self.rchild is None:
				self.rchild = Node(value, None, None)
				return
			self.rchild.insert(value)
		elif self.val > value:
			if self.lchild is None:
				self.lchild = Node(value, None, None)
				return
			self.lchild.insert(value)
		else:
			print("{0} already exists.".format(value))

	def search(self, key):
		if self.val == key:
			return True
		elif self.val > key:
			if self.lchild is None:
				return False
			return self.lchild.search(key)
		elif self.val < key:
			if self.rchild is None:
				return False
			return self.rchild.search(key)

	def display(self):
		if self.lchild is not None:
			self.lchild.display()
		
		print(self.val)
		
		if self.rchild is not None:
			self.rchild.display()


root = Node(10, None, None)
root.insert(2)
root.insert(4)
root.insert(5)
root.insert(12)
root.insert(23)
root.display()

File: Data Structures/BST/BST.cpp
#include<iostream>
using namespace std;

class node{
	int key;
	node *left;
	node *right;

	public:
	node(int value){
		key = value;
		left = NULL;
		right = NULL;
	}

	void insert(int n){
		if(n < key){
			if(left == NULL){
				left = new node(n);
			}
			left->insert(n);
		}
		if(n > key){
			if(right == NULL){
				right = new node(n);
			}
			right->insert(n);
		}
	}


	bool search(int n){
		if(n == key){
			cout << "Found...!!" << endl;
			return 1;
		}
		else if(n > key){
			if(right == NULL){
				cout << "Not Found...!!" << endl;
				return 0;
			}
			else{
				return right->search(n);
			}
		}
		else{
			if(left == NULL){
				cout << "Not Found...!!" << endl;
			}
			else{
				return left->search(n);
			}
		}
	}

	void display(){
		if(left != NULL){
			left->display();
		}
                cout << key << "  " ;
		if(right != NULL){
			right->display();
		}
	}
};


int main(){
	node *root;
	root = new node(10);
	root->insert(3); 
	root->insert(5);
	root->insert(1);
	root->display();
	root->search(3);

	return 0;
}
